### DISCUSSION TITLE
Using NL for multiclass classification task

### DISCUSSION THREAD

**Discussion by borisrakovan** [LukasZahradnik/PyNeuraLogic#27] [Q&A]
Hello,

In the past days I was trying to figure out how to use the NeuraLogic framework to imitate the computation performed by Relational graph convolutional networks (RGCN) introduced [here](https://arxiv.org/pdf/1703.06103v4.pdf). Unfortunately, I wasn't able to come up with any working solution. I just can't seem to understand how to translate such computations into the logic-based language. 

I have my examples and queries set up in a way that is very similar to the one already discussed [here](https://github.com/LukasZahradnik/PyNeuraLogic/issues/23#issue-1129953695) . I would like to perform a multiclass classification task (with 22 classes) using the following propagation rule:

![image](https://user-images.githubusercontent.com/49689325/155804505-d44d9cb6-e7c6-4e65-adef-a0418abce32e.png)

In RGCNs, the transformation applied to the neighboring nodes depends on the particular type of the relation (category) between the two nodes, and I can't figure out how to encode this using the NL syntax. I managed to come up with an initialization of the embeddings:

```
template2.add_rules([
    *[(R.person_embed(p)[3,] <= R.person(p)) for p in all_persons],
    *[(R.rel_embed(r)[3, 3] <= R.rel(r)) for r in all_relations]
])
```

where I initialize a 3x3 weight matrix for each of the relations, but I don't know whether this is the correct line of thinking. Alternatively, what would be the simplest way to set up a template for multiclass classification on this dataset - one that I could use as a starting point?

Any help would be much appreciated! Thank you.

**Comment by LukasZahradnik** [comment]
Hi, I have recently written a [short tutorial](https://pyneuralogic.readthedocs.io/en/latest/rules_mapping.html) on rules mapping that might help you understand NL syntax more.

One layer of RGCN could be rewritten as (I added comments on top of the rule to make mapping more clear; also, the multiplication is the other way around in the comment - it is actually `W * h`):

```py
t = Template()
metadata = [Aggregation.SUM, Activation.IDENTITY]

for relation in all_relations:
    #       h_i1           h_j0  *  W_r +  h_i0  *  W_0         Ensuring j \in N^r_i
    t += (R.h_1(V.I) <= (R.h_0(V.J)[1,], R.h_0(V.I)["W_0": 1,], R.hidden.edge(V.I, relation, V.J))) | metadata
    t += R.h_1 / 1 | [Activation.SIGMOID]

```

For additional layers, you would change only indices, that is:

```py
    #                                               W_0 in the 1st layer 
    t += (R.h_2(V.I) <= (R.h_1(V.J)[1,], R.h_1(V.I)["W_1": 1,], R.hidden.edge(V.I, relation, V.J))) | metadata
    t += R.h_2 / 1 | [Activation.SIGMOID]
```

You will have one rule for each relation, `R.h_1`, which computes the inner sum (the sum is defined in the attached metadata). Getting all `j`'s from `N_ir` is done automatically in the background (enforcing `j`'s to be from `N_ir` is done with `R.edge`, without it, all `j`'s would be taken into account). 

The outer sum over relations is done in the background as well - rules with the same (grounded) head are summed when queried.

The only part I omitted is the normalization constant - it's a bit tricky, and I'm not sure how you want to define it.

You could, for example, add it into the dataset examples as:

```py
    R.c("entity_1", "relation_1")[0.5]  # Some value c_{i,r}
```

And then change the template a little bit (the new (first) rule computes `c_i,r * Wr * h0j`)

```py

for relation in all_relations:
    #                                         C_i,r        * W_r * h_j0                 
    t += (R.c_0(V.I, relation, V.J) <= (R.c(V.I, relation), R.h_0(V.J)[1,])) | Metadata(activation="product-identity")
    # "product-" enforces multiplicating instead of summing body relations
    # That's why it is (C_i,r * W_r * h_j0) and not (C_i,r + W_r * h_j0)
    # the 'r' is equal to 'relation'

    t += R.c_0 / 3 | [Activation.IDENTITY]

    #      h_i1         (C_i,r * W_r * h_j0)       +  h_i0  *  W_0         Ensuring j \in N^r_i
    t += (R.h_1(V.I) <= (R.c_0(V.I, relation, V.J), R.h_0(V.I)["W_0": 1,], R.hidden.edge(V.I, relation, V.J))) | metadata
    t += R.h_1 / 1 | [Activation.SIGMOID]

```

I haven't really tested it, but it should correspond to your RGCN definition. When it comes to the multiclass classification, you should either go with a one-hot encoded vector or make a query for each possible combination of entities and relations (where you will learn query for the correct class to be 1 and others to be 0).

// EDIT: I tested the template, visualized it, and it seems to generate the correct computation graph.

**Reply by borisrakovan** [reply]
Hi, thank you very  much for your help. I read the tutorial and I feel a bit more confident in building the templates now.

However, I encountered another issue while using the provided RGCN-like template to train a model.

This is my current template:
```python
template = Template()

template += [(R.h_0(p)[1,] <= R.person(p)) for p in all_persons]

metadata = [Aggregation.SUM, Activation.IDENTITY]

for relation in all_relations:
    #       h_i1           h_j0  *  W_r +  h_i0  *  W_0         Ensuring j \in N^r_i
    template += ( R.h_1(V.I) <= (R.h_0(V.J)[1,], R.h_0(V.I)["W_0": 1,], R.hidden.edge(V.I, relation, V.J)) ) | metadata
    template += R.h_1 / 1 | [Activation.SIGMOID]

template += R.predict(V.X, V.R, V.Y)[1,] <= (R.h_1(V.X)[1,], R.h_1(V.Y)[1,])
```
For simplicity, I only decided to include 1 RGCN layer and use scalars as weights in the beginning. When I try to train a model using this template, I get the following exception:

```
java.util.InputMismatchException          Traceback (most recent call last)
[/usr/local/lib/python3.7/dist-packages/neuralogic/core/builder/builder.py](https://localhost:8080/#) in from_logic_samples(self, parsed_template, logic_samples, backend)
     57 
     58         logic_samples = java_model.s
---> 59         logic_samples = logic_samples.collect(self.collectors.toList())
     60 
     61         if backend == Backend.JAVA:

java.util.InputMismatchException: java.util.InputMismatchException: Query: predict(donald_0, aunt, dorothy_0) was not matched anywhere in the ground network - Cannot calculate its output!
```

Despite the fact that the following example-query pair is part of my dataset:

```
E:
edge(donald_0, father, michael_0)
edge(michael_0, sister, dorothy_0)
rel(aunt)
rel(father)
rel(sister)
person(dorothy_0)
person(donald_0)
person(michael_0)

Q: 
1 predict(donald_0, aunt, dorothy_0)
```

Do you know what could be the problem here?

**Reply by LukasZahradnik** [reply]
Hi, the problem is that to compute `R.predict("donald_0", "aunt", "dorothy_0")`, you have to compute `R.h_1("donald_0")` and `R.h_1("dorothy_0")` first (as defined in the rule). But the computation of  `R.h_1("dorothy_0")`  fails because there is no rule that can be satisfied - there are no edges with `dorothy_0` as the source (required by the only definition of `R.h_1`).

**Reply by borisrakovan** [reply]
Hi, that makes sense. How would I make it work? Does it make sense to add another "symmetric" propagation rule that would account for all incoming edges of a node? I remember something like that was done in the knowledge base completion task that we discussed in 
the previous issue. 

Something like:

```python
template += ( R.h_1(V.I) <= (R.h_0(V.J)[1,], R.h_0(V.I)["W_0": 1,], R.hidden.edge(V.I, relation, V.J)) ) | metadata
template += ( R.h_1(V.J) <= (R.h_0(V.I)[1,], R.h_0(V.J)["W_0": 1,], R.hidden.edge(V.I, relation, V.J)) ) | metadata
```

Also, why is the edge relation accessed through the `hidden` attribute?

Thank you.

**Reply by LukasZahradnik** [reply]
Hi, there are multiple ways how to solve it, and I can't really tell which one will suit you and your problem the most.

The way you proposed should work; maybe it is unwanted to have two different scalar weights (one per rule), but that can be resolved by naming them the same.

You could also add more edge relations to the example set, add another rule to handle nodes with no edges (e.g., as [shown here](https://pyneuralogic.readthedocs.io/en/latest/rules_mapping.html#multiple-rules)), or something else. 

You can read more about the `hidden` modifier [here](https://pyneuralogic.readthedocs.io/en/latest/modifiers.html). Basically, the same rule without the modifier:
```
R.h_1(V.I) <= (R.h_0(V.J)[1,], R.h_0(V.I)["W_0": 1,], R.edge(V.I, relation, V.J))
```
would compute

```
W * h_{0,j} + W_0 * h_{0,i} + edge_{i,relation,j}
```

The `hidden` modifier will remove the flagged relation from the computation graph (there will be no `+ edge...`) while still enforcing the relation to be satisfied. You don't have to use the modifier; it just corresponds more to RCGN with it.

**Reply by GustikS** [reply]
Hi, it's great that you've been moving forward with this together. I have just a small note on the order within the relations problem - it totally makes sense to aggregate differently in one way and the other way around (with separate parameterization) for assymetric relations:

```python
template += ( R.h_1(V.I) <= (R.h_0(V.J)[1,], R.h_0(V.I)[1,], R.hidden.edge(V.I, relation, V.J)) ) | metadata
template += ( R.h_1(V.I) <= (R.h_0(V.J)[1,], R.h_0(V.I)[1,], R.hidden.edge(V.J, relation, V.I)) ) | metadata
```
since `edge(donald_0, father, michael_0)` is clearly very different from `edge(michael_0, father, donald_0)`, while for the symmetric relations (e.g., `edge(michael_0, sister, dorothy_0)`) you can tie the weights together as suggested (or, perhaps better yet, add `edge(dorothy_0, sister, michael_0)` to the examples, as also suggested).

Or, as a simple workaround, add `R.h_1(V.I) <= (R.h_0(V.J)[1,]` to learn "default values" for unconnected entities (as also suggested by Lukas above). Best, G.

**Reply by borisrakovan** [reply]
Hi, thank you both for your suggestions, I decided to go with the 2 rules as suggested by @GustikS. However now, when I try to train the dataset in Google Colab environment, I am facing the following issue - after calling `evaluator.train(dataset)`, the `from_logic_samples()` function call takes forever. This is the whole call chain:

```
train()
set_dataset()
set_dataset()
build_dataset()
build_dataset()
build_dataset()
from_logic_samples()
```

It just gets stuck in from_logic_samples() and never actually proceeds to training iterations. This is possibly caused by the fact that for each original example query pair, I now decided to include all other 21 negative example query pairs as suggested in our previous discussion. Therefore, I tried to downsample the initial training dataset and only include ~1000 training samples, which should result in ~22000 samples in total. Is it too much computation for the framework to handle in reasonable time? Or is the issue somewhere else? 

Thanks for your replies!

**Reply by LukasZahradnik** [reply]
Hi, `from_logic_samples` calls the NeuraLogic backend, so it is stuck somewhere in Java.

How many examples do you have, and how large are they? If all of your examples are exactly the same, then you can have only one example - instead of having `n` queries and `n` examples, you can have `n` queries and `1` example. This might help if that's the case.

To know more about what is going on, you can get the output of the Java backend. You can do it this way:

```py
import sys
from neuralogic import set_stderr, set_stdout

set_stderr(sys.stderr)
set_stdout(sys.stdout)
```

If `stdout` and `stderr` do not work in google colab, you can pass in a file or any object with `write(text)` method. The output might contain some useful info.

**Reply by GustikS** [reply]
Hi,
agreed, so firstly definitely do not repeat the same example if just the query changes.

Then, it is always a good practice to test your template on some small sample, e.g. by drawing (or checking the sizes at least), to make sure you understand what it's doing - it is easy to introduce exponential complexity in a template (e.g. by variable mismatch, recursion, etc.). 
Once you have that computation (i.e. grounding/neural net creation) sorted out on a single example, the complexity will be linear in the number of examples, so you will know exactly how much time it will take in total.

And yes, definitely turn on logging from the java backend, it will show you the exact ongoing processes, sizes of the logical/neural structures, timestamps etc. (wouldn't it be better to have that on by default, Lukas?)

**Reply by borisrakovan** [reply]
Hi, I tried doing what you suggest and only include 1 example per all queries, but I am getting the following error

```
Exception: The size of examples 1000 doesn't match the size of queries 18000
```

I tried both, calling `dataset.add_query()` with list of 18 queries and with each query individually but I am getting the same error. How is it possible to only include 1 example for 18 queries if the number of queries and examples must be the same?

**Reply by LukasZahradnik** [reply]
Hi, it seems like you want to have `n` (18000) queries to `m` (1000) examples, which is not currently possible - you can only have `n` queries to `n` examples or `n` queries to `1` example. 

I was under the impression that all of your (1000) examples are the same, so then you could have only one example (instead of 1000).

If that's not the case, and your examples differ, then you can have 1000 datasets with 18 queries and 1 example each. It's not a really nice solution; maybe I will look into handling this use case in the background.

// Edit: 
Just to make it clear, when we talk about one example, it means one list of (ground) relations, such as:

```py
one_example = [R.person("a"), R.person("b")]

# add to the dataset as
dataset.add_example(one_example)
```
When we are talking about examples, it means list of list of (ground) relations, such as:
```py
examples = [
    [R.person("a"), R.person("b")],
    [R.person("c"), R.person("d")],
]

# add to the dataset as
dataset.add_examples(examples)

# or
for example in examples:
    dataset.add_example(example)
```

**Reply by GustikS** [reply]
Hi, yes I think it's really M graphs with N queries each here. The backend supports that - queries are a list for each example. We commonly tested only 1-N and N-N, so there might be some unforeseen issues, but it would make sense to add support for it here.

@borisrakovan as a "workaround" for now, you can just use the classic multi-class (vector) encoding, it's equivalent in your case, since all the queries have the same format.

**Reply by borisrakovan** [reply]
@GustikS Alright, I decided to use one-hot encoding for the relations as discussed before. 
My example-query pairs now look like this

```
QUERY
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0] predict(donald_0, dorothy_0).
EXAMPLE
edge(donald_0, father, michael_0)
edge(michael_0, sister, dorothy_0)
rel(daughter)
rel(son-in-law)
rel(niece)
rel(wife)
rel(aunt)
rel(granddaughter)
rel(nephew)
rel(uncle)
rel(brother)
rel(mother)
rel(husband)
rel(mother-in-law)
rel(father)
rel(grandfather)
rel(grandson)
rel(sister)
rel(daughter-in-law)
rel(son)
rel(grandmother)
rel(father-in-law)
person(dorothy_0)
person(michael_0)
person(donald_0)
```

And the altered template definition:

```python
template = Template()

template.add_rules(
     [(R.h_0(p)[1,] <= R.person(p)) for p in all_persons]
)

metadata = [Aggregation.SUM, Activation.IDENTITY]

for relation in all_relations:
    template += ( R.h_1(V.I) <= (R.h_0(V.J)[1,], R.h_0(V.I)[1,], R.hidden.edge(V.I, relation, V.J)) ) | metadata
    template += ( R.h_1(V.I) <= (R.h_0(V.J)[1,], R.h_0(V.I)[1,], R.hidden.edge(V.J, relation, V.I)) ) | metadata

    template += R.h_1 / 1 | [Activation.SIGMOID]

# template += R.predict(V.X, V.R, V.Y)[1,] <= (R.h_1(V.X)[1,], R.h_1(V.Y)[1,])
template += R.predict(V.X, V.Y)[22,] <= (R.h_1(V.X)[1,], R.h_1(V.Y)[1,])
```

However, I am getting the following error:
`java.lang.ArithmeticException: Incompatible dimensions for vector multiplication: [22, 1] vs [22, 1] (try transposition)`

How wouldI fix this problem in the template? It's kind of hard for me to visualize the matrix computations that happen in the background so despite trying multiple things, I can't really figure out how to set the dimensions correctly.
Thanks.

**Reply by LukasZahradnik** [reply]
Aha @GustikS , I kinda overlooked this M-N mapping; it will be fixed in the next release. The exception @borisrakovan  is experiencing now is due to a bug in the `SquaredDiff` error function.

See https://github.com/GustikS/NeuraLogic/blob/master/Algebra/src/main/java/cz/cvut/fel/ida/algebra/functions/specific/SquaredDiff.java#L22 You are multiplying `diff` with `diff` (where `diff` is a vector), but then you require that both multiplied vectors have a different orientation (https://github.com/GustikS/NeuraLogic/blob/master/Algebra/src/main/java/cz/cvut/fel/ida/algebra/values/VectorValue.java#L207). Since you are multiplying one object with itself, it can't have two different orientations.

**Reply by GustikS** [reply]
Oh, nice, so it is a bug report in the end :) I guess I never tested multi-dim MSE (regression) mode. I'll correct that shortly (or @LukasZahradnik can just include the transposition directly in the Python release already, since he found it himself).

Nevertheless you'll want to use cross-entropy instead of MSE here, since it's a classification (@LukasZahradnik that possibly brings back the debate that these flags could be induced automatically on some higher level, like in the backend).

@borisrakovan If the Java backend feels too black-box, you can either attach a debugger to it, or use a different backend, e.g. Dynet/Torch, and debug in those (Python).

**Reply by LukasZahradnik** [reply]
Hi @borisrakovan , the M:N mapping and MSE error function should be fixed in the new version (0.1.3)

**Reply by borisrakovan** [reply]
Hi, I changed the error function to cross entropy as suggested and I was able to proceed with the traning. Unfortunately, the model does not appear to be learning at all. The average loss doesn't seem to be decreasing during the training. I thought this was caused by the scalar weights, which would make sense, so I changed the sizes of the weight matrices to 3x3 (the previous KB-completion template was learning with this size). However, it didn't help. Then I thought it might be caused by the fact that I am only training on 1000 samples, which is 1/10 of the original training dataset. So I tried to increase the number of samples first to 5000, then to 3000, but anything larger than 2000 turned out to be too much for the `from_logic_samples` call to return in a reasonable amount of time :(
 
This is my current template:

```python
template.add_rules(
     [(R.h_0(p)[3,] <= R.person(p)) for p in all_persons]
)

metadata = [Aggregation.SUM, Activation.IDENTITY]

for relation in all_relations:
    #       h_i1           h_j0  *  W_r +  h_i0  *  W_0         Ensuring j \in N^r_i
    template += ( R.h_1(V.I) <= (R.h_0(V.J)[3, 3], R.h_0(V.I)["W_0": 3, 3], R.hidden.edge(V.I, relation, V.J)) ) | metadata
    template += ( R.h_1(V.I) <= (R.h_0(V.J)[3, 3], R.h_0(V.I)["W_1": 3, 3], R.hidden.edge(V.J, relation, V.I)) ) | metadata

    template += R.h_1 / 1 | [Activation.SIGMOID]

template += R.predict(V.X, V.Y)[22,] <= (R.h_1(V.X)[1, 3], R.h_1(V.Y)[1, 3])

```

And my model settings

```
settings = Settings(optimizer=Optimizer.ADAM, epochs=n_epochs, learning_rate=0.01, error_function=ErrorFunction.CROSSENTROPY)
```

And finally, this is the plot of the average loss across epochs. You can see that it's clearly not improving.

![image](https://user-images.githubusercontent.com/49689325/156941523-5ca917b0-0e39-4155-a7c2-a8b539e3bfe7.png)

Do you know why this could be?

**Reply by GustikS** [reply]
Hi, there can be multiple reasons behind this, as with any model training. The only sure thing is that increasing the number of training samples will not help you to get a lower training error. If you're not able to overfit the model on some small training sample, then there is surely something wrong with the model itself.

Quick look - you're feeding a sum of all neighbor embeddings into a sigmoid - that might not be a good idea as the sigmoid will saturate very quickly (and seize to propagate gradient). I'd put AVG there instead of SUM which is also more noisy (if you do not have some other normalization there) and generally worse. Also, tanh (or Relu) would be better than sigmoid. Overall, the model seems too linear (identity activations) except the final layer - I'd put more non-linearities. Also you can attach more weights (e.g., to the edge predicate and head of the rule h_1) - the weights can further be matrices everywhere instead of just vectors. All of these steps should increase the model capacity in order to better fit the training data. Finally, you can play with the learning rate (to prevent the error jumps, there are also lr decay schemes).

Grounding the samples with this should be pretty fast, if that's the case but you suddenly can't go over 2000 - it is possible that java start memory swapping there - increase memory to prevent that if you can.

Finally, a lot of relevant information, including whether your sigmoids are getting saturated during training, how much memory you have left, how fast/big the groundings/networks are etc. are displayed in the logs from the backend - really try to take a look at it.

If it still doesn't train after that, maybe you can share the dataset with us so we can try, too?

**Reply by GustikS** [reply]
...also, I forgot to mention classic deep learning tricks - you can, of course, increase the embedding dimensionality everywhere, and add more layers, since this single-layered model will only aggregate information from immediate neighbors, hence if you want to classify longer-range relations, you should consider stacking multiple of these.

But I would generally advise to firstly debug the model on a simple and clear sample, say 1 example and 1 couter-example, where you're sure what you're trying to achieve, check that the model does that, and then generalize to more complex relations with bigger models and sample sizes...

**Reply by borisrakovan** [reply]
Hello, so I tried to experiment with a bunch of things, including replacing SUM with AVG, sigmoids with RELUs, increasing the number of learnable parameters and adding an extra RGCN layer. I am indeed getting better results now, and the loss is also decreasing towards 0, reaching ~0.35 after 120 epochs. However, the accuracy on the training set is still only 23%. It is definitely better than random, considering there is 22 classes, but still not very good and it means that I am not even able to overfit the model on the training data. The accuracy on the test set is no better than random. I ended up with the following template. Is there anything else I could improve?

```python
template = Template()

template.add_rules(
     [(R.h_0(p)[10,] <= R.person(p)) for p in all_persons]
)

metadata = [Aggregation.AVG, Activation.IDENTITY]

# first layer
for relation in all_relations:
    template += ( R.h_1(V.I) <= (R.h_0(V.J)[10, 10], R.h_0(V.I)["W_0": 10, 10], R.hidden.edge(V.I, relation, V.J)) ) | metadata
    template += ( R.h_1(V.I) <= (R.h_0(V.J)[10, 10], R.h_0(V.I)["W_1": 10, 10], R.hidden.edge(V.J, relation, V.I)) ) | metadata

    template += R.h_1 / 1 | [Activation.RELU]

# second layer
for relation in all_relations:
    template += ( R.h_2(V.I) <= (R.h_1(V.J)[10, 10], R.h_1(V.I)["W_3": 10, 10], R.hidden.edge(V.I, relation, V.J)) ) | metadata
    template += ( R.h_2(V.I) <= (R.h_1(V.J)[10, 10], R.h_1(V.I)["W_4": 10, 10], R.hidden.edge(V.J, relation, V.I)) ) | metadata

    template += R.h_2 / 1 | [Activation.RELU]

template += R.predict(V.X, V.Y)[22,10] <= (R.h_2(V.X)[10, 10], R.h_2(V.Y)[10, 10])
```

Also, I would like to use more samples from the dataset, but the grounding process takes really long time. I tried to inspect the logs and I am getting outputs similar to this (for each example in the dataset, so if there was something other than this it would be really hard to catch):

```
...
Mar 09, 2022 4:48:00 PM cz.cvut.fel.ida.pipelines.pipes.specific.SupervisedNeuralizationPipe lambda$apply$0
INFO: Neuralizing sample s_375: [0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0]=predict(gabrielle_366, william_334) <- ex:#375 [facts:7, (in conjunctions:1, rules:0)]
Mar 09, 2022 4:48:00 PM cz.cvut.fel.ida.pipelines.pipes.specific.SupervisedNeuralizationPipe lambda$apply$3
INFO: NeuralNet created: s_375:predict(gabrielle_366, william_334): [0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0]=WeightedAtomNeuron = predict(gabrielle_366, william_334) <- net:g375, neurons: 32
Mar 09, 2022 4:48:00 PM cz.cvut.fel.ida.neural.networks.structure.transforming.LinearChainReducer reduce
INFO: LinearChainPruning reduced neurons from 32 down to 15 with prunings: 17
Mar 09, 2022 4:48:00 PM cz.cvut.fel.ida.neural.networks.structure.transforming.IsoValueNetworkCompressor reduce
INFO: IsoValue neuron compression from 15 down to 13 etalons (topologic-reconstruction: 13)
Mar 09, 2022 4:48:00 PM cz.cvut.fel.ida.pipelines.pipes.specific.SupervisedNeuralizationPipe lambda$apply$0
INFO: Neuralizing sample s_376: [0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]=predict(myrna_371, lucille_347) <- ex:#376 [facts:8, (in conjunctions:1, rules:0)]
...
```

Occasionally, I also get a log similar to this
`# WARNING: Garbage collection takes 196.77% of time!!`
Which I assume is a bug. Another possible bug I found is that once I set the std_err and std_out to the console, I am not able to turn the logging off. 

Thank you for your help.

**Reply by GustikS** [reply]
Hi, you can still attach learnable weights to the relations (edge) and rule heads (h1 and h2) and replace the identities with some non-linearity (e.g. tanh, since ReLU could still be causing linear behavior in these small networks). Or just simply remove all the metadata/activations, really. 

But all that is really just for further brute-force capacity increase. I'd strongly suggest to avoid these black-box trials and carefully inspect the model structure instead (as suggested above). Particularly, inspect whether the information necessary to distinguish the samples does not get lost somewhere during the inference (causing the random behavior). Personally, I'd start with just taking 2 samples you know exactly how to distinguish (e.g. manually crafted), inspecting the model inference (via visualization or debugger), and try to e.g. manually set-up the weights so that you confirm the model has the capacity to correctly classify these (and see why!). Then try to learn these weights automatically (and, potentially, see why not). Then add some more, principally different sample (e.g. more relations, further apart, etc.), and repeat. Once this gets smooth across different sample types, you can move to the black-box tuning mode on the full dataset.

Again, if you don't feel comfortable debugging the model like that, you can use the Java backend just as a grounder ("neural model creator"), export to Dynet/Torch, and continue debugging from there (or even manually create the models from scratch there, once you understand the model principle/structure). This could be also good to confirm the alignment, and potential discrepancies, from your reference RGCN implementation.

For instance, one thing I notice that might be different from the RGCN formula is that the W_0 * h_i is outside of the sums (i.e. you're not adding that for each relation and each neighbor), so you might want to move that into a separate rule (no need to name the weight):

```python
R.h_1(V.I) <= R.h_0(V.I)[10,10]
```

Regarding the logs - you can filter it via standard logging levels. But it seems that your logging level is set to INFO only. @LukasZahradnik did you limit that? If so, can you give option to turn that to at least FINE or FINER - that would include more useful info about the grounding here (would be good to see the Herbrand sizes etc.). Also, on my system it shows time deltas in milliseconds (that would be more helpful) - @LukasZahradnik, did you turn that off somehow? :) Perhaps it's the Colab formatter - I'd suggest running locally for this. 

From the warning, it seems that the GC is throttling the process (this means that you've got eventually completely blocked by GC on 2 threads), which might explain why the grounding gets stuck around the 2000 samples (as I speculated above, or is it slow from the start?). To avoid this you can increase the Java virtual memory by passing e.g. -Xmx4g to the JVM (default is 512m I think). But if the grounding is generally slow from the start, there will be some problem with the model instead (which will be seen from the detailed logs, too).

I wish you strong nerves :)

**Reply by LukasZahradnik** [reply]
Hi, 

@GustikS I'm not limiting anything. I'm only setting the Java's `stdout` and `stderr`, which I assumed would be enough. It seems like I will have to work directly with loggers instead, which would also resolve (re)setting `stdout/stderr` problem @borisrakovan  mentioned. There might be some issues because each class has its own logger, but we will see...


> For instance, one thing I notice that might be different from the RGCN formula is that the W_0 * h_i is outside of the sums

This is kinda my bad. I overlooked it while rewriting it into PyNeuraLogic. Regarding the GC issue, maybe it might make sense to trigger GC manually from time to time in the backend/frontend?


@borisrakovan if you choose to go with DyNet/Torch, I would recommend you to go with Torch, as PyNeuraLogic has implemented better support for it (correct broadcasting etc.).
Also, increasing the memory is not possible in the latest release. I have it [implemented](https://github.com/LukasZahradnik/PyNeuraLogic/blob/master/neuralogic/__init__.py#L26) and will release it soon with some other changes.

**Reply by LukasZahradnik** [reply]
Hello, I fixed the logging in the latest version (0.1.4). The usage is following:

```py
import sys

from neuralogic.logging import add_handlers, clear_handlers


add_handlers(sys.stdout)  # by default FINER and COLOR Formatter

clear_handlers()  # Clear ALL handlers
```

**Reply by borisrakovan** [reply]
Hi @GustikS @LukasZahradnik ,
 I am trying to use the Torch backend as suggested, but it gives me `NotImplementedError`. From the code it looks like the crossentropy loss is indeed not implemented:

```python
if error_function not in TorchEvaluator.error_functions:
    raise NotImplementedError
```

Also, as a side note, the grounding process is faster on my local machine then it is in Colab - it only takes about ~2 minutes for 1600 samples, which is still not ideal, but it's much better than ~10 minutes in Colab.

**Reply by LukasZahradnik** [reply]
Hi @borisrakovan 

You can add the Crossentropy error function yourself by adding it into the dictionary [here](https://github.com/LukasZahradnik/PyNeuraLogic/blob/master/neuralogic/nn/evaluators/torch.py#L18) (just import the class and add it into the dictionary), or you can define your own training loop. I will try to add it to one of the next releases.

I'm checking out your problem, and it is indeed slow. I will debug the model and let you know later on.
